### PR TITLE
[WIP]Add keyboard shortcuts to file browser.

### PR DESCRIPTION
**Pull request by niole**
Resolves issue #215 

Notes to reviewers:
 * Please help suggest strategies for refactoring code to be consistent to notebook keyboard shortcuts (see `keyboardmanager.js` and `actions.js`)

### CODE REVIEW

File: notebook/static/tree/js/main.js [hunk 1]
```diff
@@ -37,6 +37,7 @@ require([
     loginwidget){
     "use strict";
 
+    //gets notebooklist creates a method
```
Comment (minrk) [inline notebook/static/tree/js/main.js]: this comment isn't accurate

Comment (niole) [inline notebook/static/tree/js/main.js]: Thanks for the feedback. Any input on the jquery I wrote and how to
refactor it?

File: notebook/static/base/js/namespace.js [hunk 2]
```diff
@@ -32,9 +32,9 @@ define(function(){
     // expose modules
     
     jprop('utils','base/js/utils')
-    
+  
     //Jupyter.load_extensions = Jupyter.utils.load_extensions;
-    //
```
Comment (minrk) [inline notebook/static/base/js/namespace.js]: Looks like you've got whitespace stripping turned on, and commits to files you didn't actually change. We would prefer to avoid committing that sort of change.

**Comment by minrk** [comment]
We have utilities for dealing with keyboard events already that we use in the notebook, it probably doesn't make sense to write custom keyboard handling like this, with hardcoded integer keycodes, etc.

Anywhere we are going to add new keyboard shortcuts should possibly be doing it through the actions and keyboard manager stuff we set up for the notebook.

**Comment by niole** [comment]
Thanks for the feedback. I'm going to take a look around in the code and
try to start moving things around.

**Comment by niole** [comment]
Do you think it would be better to:

 1. Create new `keyboardmanager.js`, `keyboard.js`, and `actions.js` files
for the directory tree or
 2. Implement into the existing keyboard handling utilities, using the URL
(or maybe something else e.g. event methods) to determine whether we're in
a notebook or directory tree view

I think that 1. would mean less irrelevant code gets loaded to the client
but may introduce more code duplication than 2.

**Comment by niole** [comment]
Do you think it would be better to:

 1. Create new `keyboardmanager.js`, `keyboard.js`, and `actions.js` files
for the directory tree or
 2. Implement into the existing keyboard handling utilities, using the URL
(or maybe something else e.g. event methods) to determine whether we're in
a notebook or directory tree view

I think that 1. would mean less irrelevant code gets loaded to the client
but may introduce more code duplication than 2.

**Comment by niole** [comment]
resolves issue #215

File: notebook/Untitled.ipynb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (feynmanliang) [inline notebook/Untitled.ipynb]: Delete this file

File: notebook/Untitled1.ipynb [hunk 4]
```diff
@@ -0,0 +1,65 @@
+{
```
Comment (feynmanliang) [inline notebook/Untitled1.ipynb]: Delete this file

File: notebook/base/Untitled.ipynb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (feynmanliang) [inline notebook/base/Untitled.ipynb]: Delete this file

File: notebook/static/notebook/js/keyboardmanager.js [hunk 6]
```diff
@@ -25,24 +25,32 @@ define([
...
-        this.command_shortcuts.add_shortcuts(this.get_default_common_shortcuts());
-        this.command_shortcuts.add_shortcuts(this.get_default_command_shortcuts());
-        this.edit_shortcuts = new keyboard.ShortcutManager(undefined, options.events, this.actions, this.env);
-        this.edit_shortcuts.add_shortcuts(this.get_default_common_shortcuts());
-        this.edit_shortcuts.add_shortcuts(this.get_default_edit_shortcuts());
+        if (this.view === 'tree'){
+          this.tree_shortcuts = new keyboard.ShortcutManager(undefined, options.events, this.actions, this.env);
+          this.tree_shortcuts.add_shortcuts(this.get_default_tree_shortcuts());
+        }
+        if (this.view === 'notebook'){
```
Comment (feynmanliang) [inline notebook/static/notebook/js/keyboardmanager.js]: `else if`

Comment (feynmanliang) [inline notebook/static/notebook/js/keyboardmanager.js]: Actually since you set `this.view = 'tree'` on L32, this branch will never execute. Can you test to make sure that notebook view shortcuts have not regressed?

File: notebook/static/notebook/js/keyboardmanager.js [hunk 7]
```diff
@@ -60,6 +68,16 @@ define([
      * // e.g. 'shift' , 'shift-enter', 'cmd-t'
      *```
      */
+
+    KeyboardManager.prototype.get_default_tree_shortcuts = function() {
+        return {
+            'enter'       : 'ipython.enter-row',
```
Comment (feynmanliang) [inline notebook/static/notebook/js/keyboardmanager.js]: Align on `:`

File: notebook/static/notebook/js/keyboardmanager.js (hunk 6 above)
Comment (niole) [inline notebook/static/notebook/js/keyboardmanager.js]: the way things update, this.view updates the way you want it to. you won't
run into problems

File: notebook/static/notebook/js/keyboardmanager.js [hunk 8]
```diff
@@ -25,24 +25,32 @@ define([
          * @class KeyboardManager
          * @constructor
          * @param options {dict} Dictionary of keyword arguments :
-         *    @param options.events {$(Events)} instance 
+         *    @param options.events {$(Events)} instance
          *    @param options.pager: {Pager}  pager instance
          */
         this.mode = 'command';
+        this.view = 'tree';
```
Comment (feynmanliang) [inline notebook/static/notebook/js/keyboardmanager.js]: Move this into `KeyboardManager.prototype.bind_events()`

File: notebook/static/notebook/js/keyboardmanager.js [hunk 9]
```diff
@@ -123,7 +141,14 @@ define([
 
     KeyboardManager.prototype.bind_events = function () {
         var that = this;
+        var notebook_view = window.document.getElementById("pager");
```
Comment (feynmanliang) [inline notebook/static/notebook/js/keyboardmanager.js]: This doesn't seem to be a very robust way to test whether we are in notebook or tree view (a change to the HTML will break this logic). Is there a better way to test for this?

If not, please add a comment that this test expects "pager" to only exists when in notebook view.

File: notebook/static/notebook/js/keyboardmanager.js [hunk 10]
```diff
@@ -145,25 +175,27 @@ define([
...
-        }
-        
-        if (this.mode === 'edit') {
-            return this.edit_shortcuts.call_handler(event);
-        } else if (this.mode === 'command') {
-            return this.command_shortcuts.call_handler(event);
+
+        if (this.view === 'tree'){
+          this.tree_shortcuts.call_handler(event);
+        }else{
```
Comment (feynmanliang) [inline notebook/static/notebook/js/keyboardmanager.js]: space between `}` and `else`

File: notebook/static/notebook/js/keyboardmanager.js [hunk 11]
```diff
@@ -145,25 +175,27 @@ define([
...
+          if (!this.enabled) {
+              if (event.which === keycodes.esc) {
+                  this.notebook.command_mode();
+                  return false;
+              }
+              return true;
+          }
+          if (this.mode === 'edit') {
```
Comment (feynmanliang) [inline notebook/static/notebook/js/keyboardmanager.js]: This logic is only present in notebook view (tree view has no modes); move into the else branch on the `this.view === 'tree'` conditional

File: notebook/static/tree/js/main.js [hunk 12]
```diff
@@ -54,18 +60,32 @@ require([
     var session_list = new sesssionlist.SesssionList($.extend({
         events: events},
         common_options));
+
     var contents = new contents_service.Contents({
         base_url: common_options.base_url,
         common_config: common_config
     });
+
+    page = new page.Page();
```
Comment (feynmanliang) [inline notebook/static/tree/js/main.js]: Move back to L42; in general, minimize unnecessary changes to codebase

File: notebook/static/tree/js/notebooklist.js [hunk 13]
```diff
@@ -315,6 +316,35 @@ define([
         this._selection_changed();  
     };
 
+    NotebookList.prototype.is_selected = function(){
+      if ($('.focus').size() === 0){
+        $('.list_item').first().attr("tabindex","0").addClass('focus').focus();
+      }else{
```
Comment (feynmanliang) [inline notebook/static/tree/js/notebooklist.js]: space before `else`

File: notebook/static/tree/js/notebooklist.js [hunk 14]
```diff
@@ -315,6 +316,35 @@ define([
         this._selection_changed();  
     };
 
+    NotebookList.prototype.is_selected = function(){
```
Comment (feynmanliang) [inline notebook/static/tree/js/notebooklist.js]: This isn't really `is_selected` since there is a side effect to focus on the first list item if nothing is in focus. Where is this function used?

File: notebook/static/tree/js/treeactions.js [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+                }
+                if (focused === 0){
+                  //focus first elt in .list_items
+                  env.notebooklist.select_row(0);
+                }
+            }
+        }
+    };
+
+    /**
```
Comment (feynmanliang) [inline notebook/static/tree/js/treeactions.js]: Are any of these actions used in NotebookList? If not, remove from this file.

File: notebook/templates/notebook.html [hunk 16]
```diff
@@ -324,6 +324,6 @@
 
 <script src="{{ static_url("components/text-encoding/lib/encoding.js") }}" charset="utf-8"></script>
 
-<script src="{{ static_url("notebook/js/main.min.js") }}" charset="utf-8"></script>
+<script src="{{ static_url("notebook/js/main.js") }}" charset="utf-8"></script>
```
Comment (feynmanliang) [inline notebook/templates/notebook.html]: Point towards minified version

**Comment by feynmanliang** [comment]
Please include a summary of the changes (e.g. added `tree_shortcuts` to `KeyboardManager`, copied `actions.js` -> `treeactions.js` and made modifications XYZ, etc). Also state the default bindings provided in `get_default_tree_shortcuts`.

File: notebook/static/notebook/js/keyboardmanager.js (hunk 6 above)
Comment (feynmanliang) [inline notebook/static/notebook/js/keyboardmanager.js]: I see now, it's set inside `bind_events()`. I think that L32 belongs inside `bind_events()` to keep logic centralized

**Comment by niole** [comment]
@minrk @takluyver please review.
Overview of changes/additions:

  In `tree/js/main.js`, initialize new `treeactions.js` file via `actions.init()`, pass into a `KeyboardManager` instance, which is then passed into a `NotebookList` instance. The new `notebook_list` object is then added to the `env` local to `ActionHandler` in `treeaction.js`, via a `.set_notebooklist` function I added to the `KeyboardManager` prototype.
  Added new functions to the `NotebookList` prototype to handle keyboard events:
`.get_selected_index`
`.get_length_siblings`
`.select_row`
`.open`
  Created  the actions file for instances of notebooklist, `tree/js/treeactions.js`, which is a copy of `notebook/js/actions.js`, but has different enters in the `_actions` object.
  Created a set of default shortcuts for the file tree by adding `.get_default_tree_shortcuts` to the `KeyboardManager` prototype, which returns action names for 'enter', 'j', and 'k'.
  Also, added a `view` variable to the `KeyboardManager` prototype in order to differentiate when to build shortcuts for the file tree or for a newly created notebook. This variable is updated in `KeyboardManager.prototype.bind_events`.

File: notebook/static/notebook/js/keyboardmanager.js (hunk 8 above)
Comment (takluyver) [inline notebook/static/notebook/js/keyboardmanager.js]: Actually, I'd rather rethink this to produce a cleaner architecture.

There should be a lower-level keyboard piece of code that handles the basic architecture of listening for shortcuts and responding to them, and then the notebook and the tree view should be able to build on that separately to listen for the shortcuts they want. You may have to split up and refactor some of the code to do this.

However, there's some big refactorings going on for 5.0, and I don't know if this code will still be used in its present form. @Carreau , @jdfreder : do either of you know what the fate of the keyboard interaction code will be?

File: notebook/static/tree/js/treeactions.js [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+     *
+     *  All action provided by Jupyter are prefixed with `ipython.`.
+     *
+     *  One can register extra actions or replace an existing action with another one is possible
+     *  but is considered undefined behavior.
+     *
+     **/
+    var _actions = {
+        'enter-row': {
+            help    : 'enters row',
```
Comment (takluyver) [inline notebook/static/tree/js/treeactions.js]: I'm not sure what this means. 'Opens current file'?

File: notebook/static/tree/js/treeactions.js [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+            help_index : 'aa',
+            handler : function (env) {
+                var len_sib = env.notebooklist.get_length_siblings('.focus');
+                if (len_sib > 0){
+                  env.notebooklist.open();
+                }
+            }
+        },
+        'select-previous-row':{
+            help    : 'selects previous row',
```
Comment (takluyver) [inline notebook/static/tree/js/treeactions.js]: On the notebook list, 'select' to me means checking the checkbox next to an item. I'd describe what this does in terms of a 'cursor' or 'focus', e.g. 'move cursor down'.

File: notebook/templates/tree.html [hunk 19]
```diff
@@ -167,5 +167,5 @@
 {% block script %}
     {{super()}}
 
-<script src="{{ static_url("tree/js/main.min.js") }}" type="text/javascript" charset="utf-8"></script>
+<script src="{{ static_url("tree/js/main.js") }}" type="text/javascript" charset="utf-8"></script>
```
Comment (takluyver) [inline notebook/templates/tree.html]: It's fine to change this for testing, but please don't commit the change.

**Comment by takluyver** [comment]
As @minrk and @feynmanliang have pointed out, there's quite a few unwanted changes that have got into this PR - files added and whitespace-only changes. We'll need to rebase this to trim those changes out.

**Comment by niole** [comment]
* Thanks for the requested changes. I'm going to start working on it for
the sake of cleaning up the code, but I'm curious to hear about the
refactorings for the 5.0 milestone/if this PR bears significance therein.
* the `'enter-row'` action name in `_actions`  was made to open the focused
file/directory on enter's keydown
* I'm also curious to know if anyone has an opinion on `this.view` in
`KeyboardManager`, as in should it be initially defined in the
`KeyboardManager` constructor or in
`KeyboardManager.prototype.bind_events`, which is where it gets assigned.
*  Also wondering if anyone has an opinion on the location of the `if
(this.view === 'tree')` statement on line 186 in `keyboardmanager.js`,

**Comment by takluyver** [comment]
See my inline comment - I think it would be best to refactor things a bit so the keyboard manager doesn't have a `this.view` or conditional logic based on it.

**Comment by niole** [comment]
Gotcha

File: notebook/static/notebook/js/keyboardmanager.js (hunk 8 above)
Comment (jdfreder) [inline notebook/static/notebook/js/keyboardmanager.js]: The keyboard interaction will change I think, but in time.  I think while we are still maintaining this branch, we might as well consider allowing this feature to be merged.  I'm of a similar mindset to @Carreau and am all for the addition of his and @svurens command palette, despite the refactor that is happening.

**Comment by niole** [comment]
@takluyver @mink reset unnecessary commits and, renamed action names in treeactions.js to be  more descriptive, removed unused action handlers in notebooklist.js, moved this.view into KeyboardManager.bind_events to help clarify it's use case

**Comment by niole** [comment]
@takluyver @mink Should I close this?

**Comment by rgbkrk** [comment]
Min and Thomas have been transitioning to the other side of the globe, I'd just wait a bit for them to get back.

**Comment by minrk** [comment]
@niole there's still quite a bit of work to do on this PR to get it into working shape, including addressing the comments about avoiding committing the extra unintended changes. If you aren't prepared to finish the work, we can close it here, for now. You are always free to try again another time, but I would recommend starting with something smaller. Keyboard shortcuts in browsers are one of the more complicated and hairy aspects of the notebook.

**Comment by niole** [comment]
@mink I rebased and squashed the commit history. let me know how it's looking.

**Comment by minrk** [comment]
@niole sorry, I don't think this has been rebased correctly. It now has many commits from others, going back months. Perhaps you rebased it on the wrong branch? The commit it should be based on (as of right now) is d43d72f10d8a425432351222129d9100e64e8827.

**Comment by niole** [comment]
@minrk you're right, I should've checked twice. I think it's fixed now!

**Comment by Carreau** [comment]
This would need a rebase (and removing WIP from title). 
It seem it has been forgotten as it slipped on the second page of PR list. Sorry about that.

**Comment by niole** [comment]
@Carreau , I recall that when I put this up, Jupyter was considering a widespread refactor of the entire app. I can't imagine a rebase would bring this into a mergeable staging. Closing for now.

**Comment by Carreau** [comment]
> @Carreau , I recall that when I put this up, Jupyter was considering a widespread refactor of the entire app. I can't imagine a rebase would bring this into a mergeable state. Closing for now.

ok, I'll try to have a look at how hard a rebase would be !
Sorry for that.
